Modules/socketmodule.c

```diff
@@ -3332,32 +3332,55 @@ sock_setsockopt(PyObject *self, PyObject *args)
 {
     PySocketSockObject *s = _PySocketSockObject_CAST(self);
 
+    Py_ssize_t arglen;
     int level;
     int optname;
     int res;
-    Py_buffer optval;
+    Py_buffer buffer;
     int flag;
     unsigned int optlen;
-    PyObject *none;
+    PyObject *optval;
+
+    arglen = PyTuple_Size(args);
+    switch (arglen) {
+        case 3:
+            if (!PyArg_ParseTuple(args, "iiO:setsockopt",
+                                  &level, &optname, &optval)) {
+                return NULL;
+            }
```

Comment:
Can you please move this test for None and 4 arguments here?

`if (optval == Py_None) { PyErr_Format(...) }`